src/Features/LanguageServer/Protocol/RequestDispatcher.cs

```diff
@@ -9,91 +9,126 @@
+            IEnumerable<Lazy<AbstractRequestHandlerProvider, RequestHandlerProviderMetadataView>> requestHandlerProviders, WellKnownLspServerKinds serverKind)
         {
-            var requestHandlerDictionary = ImmutableDictionary.CreateBuilder<string, Lazy<IRequestHandler>>(StringComparer.OrdinalIgnoreCase);
+            var requestHandlerDictionary = ImmutableDictionary.CreateBuilder<RequestHandlerMetadata, Lazy<IRequestHandler>>();
 
             // Store the request handlers in a dictionary from request name to handler instance.
             foreach (var handlerProvider in requestHandlerProviders)
             {
-                var methods = handlerProvider.Metadata.Methods;
+                var handlerTypes = handlerProvider.Metadata.HandlerTypes;
                 // Instantiate all the providers as one lazy object and re-use it for all methods that the provider provides handlers for.
                 // This ensures 2 things:
                 // 1.  That the handler provider is not instantiated (and therefore its dependencies are not) until a handler it provides is needed.
                 // 2.  That the handler provider's CreateRequestHandlers is only called once and always returns the same handler instances.
-                var lazyProviders = new Lazy<ImmutableDictionary<string, IRequestHandler>>(() => handlerProvider.Value.CreateRequestHandlers(serverKind).ToImmutableDictionary(p => p.Method, p => p, StringComparer.OrdinalIgnoreCase));
+                var lazyProviders = new Lazy<ImmutableDictionary<string, IRequestHandler>>(() => handlerProvider.Value.CreateRequestHandlers(serverKind)
+                    .ToImmutableDictionary(p => GetRequestHandlerMethod(p.GetType()), p => p, StringComparer.OrdinalIgnoreCase));
 
-                foreach (var method in methods)
+                foreach (var handlerType in handlerTypes)
                 {
+                    var (requestType, responseType) = ConvertHandlerTypeToRequestResponseTypes(handlerType);
+                    var method = GetRequestHandlerMethod(handlerType);
+
                     // Using the lazy set of handlers, create a lazy instance that will resolve the set of handlers for the provider
                     // and then lookup the correct handler for the specified method.
-                    requestHandlerDictionary.Add(method, new Lazy<IRequestHandler>(() => lazyProviders.Value[method]));
+                    requestHandlerDictionary.Add(new RequestHandlerMetadata(method, requestType, responseType), new Lazy<IRequestHandler>(() => lazyProviders.Value[method]));
                 }
             }
 
             return requestHandlerDictionary.ToImmutable();
+
+            static string GetRequestHandlerMethod(Type handlerType)
+            {
+                // Get the LSP method name from the handler's method name attribute.
+                var methodAttribute = Attribute.GetCustomAttribute(handlerType, typeof(MethodAttribute)) as MethodAttribute;
```

Comment:
```suggestion
                var methodAttribute = (MethodAttribute?)Attribute.GetCustomAttribute(handlerType, typeof(MethodAttribute));
```

That way any type mismatch is a useful throw instead of this which might be more confusing.

Reply:
this gives me nullable warnings because GetCustomAttribute returns a nullable Attribute